src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs

```diff
@@ -82,21 +82,33 @@ public static (Stream stream, bool usesTranscodingStream) GetStream(Stream inner
         }
     }
 
-    public static async ValueTask SendErrorResponse(HttpResponse response, Encoding encoding, Status status, JsonSerializerOptions options)
+    public static async ValueTask SendErrorResponse(HttpResponse response, Encoding encoding, Status status, Metadata? trailers, JsonSerializerOptions options)
     {
         if (!response.HasStarted)
         {
             response.StatusCode = MapStatusCodeToHttpStatus(status.StatusCode);
             response.ContentType = MediaType.ReplaceEncoding("application/json", encoding);
         }
 
-        var e = new Google.Rpc.Status
+        var e = GetStatusDetails(trailers) ?? new Google.Rpc.Status
         {
             Message = status.Detail,
             Code = (int)status.StatusCode
         };
 
         await WriteResponseMessage(response, encoding, e, options, CancellationToken.None);
+
+        static Google.Rpc.Status? GetStatusDetails(Metadata? trailers)
+        {
+            const string GrpcStatusDetailsHeader = "grpc-status-details-bin";
+            var statusDetails = trailers?.Get(GrpcStatusDetailsHeader);
+            if (statusDetails != null && statusDetails.IsBinary)
+            {
+                return Google.Rpc.Status.Parser.ParseFrom(statusDetails.ValueBytes);
```

Comment:
Make it more obvious what has gone wrong if there is invalid content in `grpc-status-details-bin`.
```suggestion
                try
                {
                    return Google.Rpc.Status.Parser.ParseFrom(statusDetails.ValueBytes);
                }
                catch (Exception ex)
                {
                    throw new InvalidOperationException(@"Error when parsing the ""grpc-status-details-bin"" trailer.", ex);
                }
```